File: C++/inorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//There are two different methods
//Method 1:  using the recursive method, repeat calling itself
class Solution {
public:
    void inorder(TreeNode* root, std::vector<int>& result)
    {
        //This is our base case
        if(!root) //similar as root == NULL
        {
            return;
        }
        inorder(root->left, result);
        result.push_back(root->val);
        inorder(root->right, result);
    }

    vector<int> inorderTraversal(TreeNode* root) {

        std::vector<int> result;
        inorder(root, result);
        return result;
    }
};

/*
//Method 2: using the iteration method, stack -> LIFO to push and pop the values
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> s;
        TreeNode *current = root;

       while (current || !s.empty()) 
       {
            while (current)
            {
                s.push(current);
                current = current->left;
            }

            current = s.top(); //access the element without remove it
            s.pop(); // remove the element
            result.push_back(current->val);

            current = current->right;
        }
        return result;
    }
};






*/

File: python/inorderTraversal.py
https://leetcode.com/problems/binary-tree-inorder-traversal/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#There are two different methods
#Method 1:  using the recursive method, repeat calling itself
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []

        def inorder(root):
            #this is our base case
            if not root:
                return
            inorder(root.left)
            result.append(root.val)
            inorder(root.right)

        inorder(root)
        return result


#Method 2: using the iteration method, stack -> LIFO to push and pop the values
"""
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        stack = []    
        current = root

        #current pointer and stack are not empty
        while current or stack:
            while current:
                stack.append(current)
                current = current.left #move our current pointer down to the left
            #until the current is pointing to the NULL, exit the while loop
            current = stack.pop()
            result.append(current.val)
            #after append it, then shift to the right
            current = current.right
        return result
"""
